patches/tModLoader/Terraria/ModLoader/DamageClass.cs

```diff
@@ -1,35 +1,48 @@
-﻿using Terraria.Localization;
+﻿#nullable enable
+using Newtonsoft.Json.Serialization;
+using System.Collections.Generic;
+using Terraria.Localization;
 
 namespace Terraria.ModLoader
 {
 	public abstract class DamageClass : ModType
 	{
+		public static Generic Generic { get; private set; } = new Generic();
+		public static NoScaling NoScaling { get; private set; } = new NoScaling();
 		public static Melee Melee { get; private set; } = new Melee();
 		public static Ranged Ranged { get; private set; } = new Ranged();
 		public static Magic Magic { get; private set; } = new Magic();
 		public static Summon Summon { get; private set; } = new Summon();
 		public static Throwing Throwing { get; private set; } = new Throwing();
 
-		internal int index;
-
 		/// <summary> This is the translation that is used behind <see cref="DisplayName"/>. The translation will show up when an item tooltip displays 'X [ClassName]'. This should include the 'damage' part. </summary>
 		public ModTranslation ClassName { get; internal set; }
 
 		/// <summary> This is the name that will show up when an item tooltip displays 'X [ClassName]'. This should include the 'damage' part. </summary>
 		public string DisplayName => DisplayNameInternal;
+		/// <summary>
+		/// This defines the classes that this DamageClass will benefit from (other than itself) for the purposes of stat bonuses, such as damage and crit chance.
+		/// Returns null by default, which does not let any other classes boost this DamageClass.
+		/// </summary>
+		public virtual Dictionary<DamageClass, float>? BenefitsFrom() {
+			return null;
+		}
```

Comment:
This is also too slow as something which creates a dictionary every time it's called. It's probably also worth making it work with crit and knockback independently. 

Also, because of how `*` works with `Modifier`, returning any float except `1` won't do what the modder expects.

I don't think this is a bad idea, but it might be better to do something like `Get[Damage/Crit/Knockback]Modifier(`Player player`)` with examples on how to combine them properly (maybe some helper methods)

By default it would just return `player.GetDamage/Crit/Knockback(this)`

Reply:
ah. I was under the impression ``Modifier`` actually treats ``*`` properly. good to hear it does not

(out of curiosity, what exactly are you suggesting I do? not sure I understand your prompt here)

Reply:
Continuing off what CB said, dictionaries aren't worth it unless you are caching, and even then for most use cases we will have here (hybrid classes benefiting from 2-3 damage classes) the performance from dictionary look up is not worth it. It would be faster to just use direct comparison.

Perhaps something like this would be better.
```cs
public virtual float BenefitsFrom(DamageClass damageClass) => 0;
```
Then either use a property or an interface containing the method to check if it is overridden.

This would eliminate the constant creation of dictionaries, but it still isn't the best.

Reply:
that's a really good idea actually

so I did that and I also made a similar style of change to CountsAs